cli/napi/env.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+
+use deno_runtime::deno_napi::*;
+
+/// # Safety
+///
+/// It's an N-API symbol
+#[no_mangle]
+pub unsafe extern "C" fn napi_fatal_error(
+  location: *const c_char,
+  location_len: isize,
+  message: *const c_char,
+  message_len: isize,
+) -> ! {
+  let location = if location.is_null() {
+    None
+  } else {
+    Some(if location_len < 0 {
+      std::ffi::CStr::from_ptr(location).to_str().unwrap()
+    } else {
+      let slice = std::slice::from_raw_parts(
+        location as *const u8,
+        location_len as usize,
+      );
+      std::str::from_utf8(slice).unwrap()
+    })
+  };
+  let message = if message_len < 0 {
+    std::ffi::CStr::from_ptr(message).to_str().unwrap()
+  } else {
+    let slice =
+      std::slice::from_raw_parts(message as *const u8, message_len as usize);
+    std::str::from_utf8(slice).unwrap()
+  };
+  panic!(
+    "Fatal exception triggered by napi_fatal_error!\nLocation: {:?}\n{}",
+    location, message
+  );
+}
+
+#[napi_sym::napi_sym]
+fn napi_fatal_exception(env: *mut Env, value: napi_value) -> Result {
+  let env: &mut Env = env.as_mut().ok_or(Error::InvalidArg)?;
+  let value = transmute::<napi_value, v8::Local<v8::Value>>(value);
+  let error = value.to_rust_string_lossy(&mut env.scope());
+  panic!(
+    "Fatal exception triggered by napi_fatal_exception!\n{}",
+    error
+  );
+}
+
+// TODO: properly implement
+#[napi_sym::napi_sym]
+fn napi_add_env_cleanup_hook(
```

Comment:
We don't have cleanup hooks in Deno. Leaving unimplemented for the first pass.

Reply:
What are these cleanup hooks supposed to do?

Reply:
Callbacks that run when exiting Deno.

Reply:
Actually in node.js, I don't think it is called when exiting the main process, rather it is only called when child worker threads exit. Node loads the target NAPI library once, and if/when child threads try to load the same NAPI library, the same library (loaded symbols) is reused, and the module initialization method is called (again) for that worker thread. So `napi_add_env_cleanup_hook` exists so that when a worker thread exits, the NAPI code can clean up resources from that worker and still continue to properly operate with the remaining worker threads.

And yeah, I was a little surprised by this behavior as well :).

Reply:
Oh, thanks for explanation @kriszyp, this is somewhat surprising. @littledivy does your implementation behave in similar way? Ie. reusing the symbols for worker threads?